### PR TITLE
Core: Add args feature

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: #9811 

## What I did

Added a new feature to stories called "args"

The idea of this feature is to allow addons to control rendering of stories via "inputs" (and capture "outputs"). Args are initialized by type inference (when available, eg. from TypeScript in your stories, or PropTypes on your components), or by user configuration (via `parameters`) and can be controlled in various ways in the Storybook UI.

Args map to the rows in the props table in docs mode.

Basic DX of using args:

```js
export const MyStory = ({ primary, size }) => <MyComponent primary={primary} size={size} />;
```

### More info:

  - See the README in this PR

## Naming

_There has been a bit of back and forth on names and I thought it was useful to outline why we settled on "args" in this PR._ 

We wanted to capture the idea of "arguments to the story function". It is worth considering that in different frameworks there can be different "categories" of arguments of your component.

 - Although in **React** there is only "props", capturing both inputs and outputs from the component;
 - in **Angular**, there are "properties" and "events"
 - in **Vue**, there are "props" and "events", and "slots", etc etc [apologies if I have details wrong here].

So, although we considered "inputs" it felt like it only captured one side of the picture.

"Properties" seemed a bit long, and "props" seemed a bit React-specific.

Another option considered was "controls" but that seemed too specific to the use case of wanting to change the value of these things via a UI element, which isn't the whole story here.

## Sub features.
_There are a few things in this PR that enable follow on features built using args:_

 - [x] Parameter enhancers -- allow an addon to generate parameters for a story at runtime. This allows complex configuration of arg "types" to be built up from the component type definitions, amongst other things

 - [x] `parameters.passArgsFirst` -- a new setting (currently defaulting to `false`, although that may change in 6.0 based on feedback), which means args are seperated from the larger story context and passed first into stories.

 - [x] `useArgs()` -- a simple hooks API available on client and server for addons to read and set arg values.

### CODE REVIEW

File: lib/client-api/README.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+The story store contains the list of stories loaded in a Storybook.
+
+Each story is loaded via the `.add()` API and contains the follow attributes, which are available on the `context` (which is passed to the story's render function and decorators):
+
+- `kind` - the grouping of the story, typically corresponds to a single component. Also known as the `title` in CSF.
+- `name` - the name of the specific story.
+- `id` - an unique, URL sanitized identifier for the story, created from the `kind` and `name`.
+- `parameters` - static data about the story, see below.
+- `args` - dynamic inputs to the story, see blow.
```
Comment (wKich) [inline lib/client-api/README.md]:
```suggestion
- `args` - dynamic inputs to the story, see below.
```

Comment (tmeasday) [reply lib/client-api/README.md]: Thanks!

File: lib/api/src/index.tsx [hunk 2]
```diff
@@ -423,3 +427,13 @@ export function useStoryState<S>(defaultState?: S) {
   const { storyId } = useStorybookState();
   return useSharedState<S>(`story-state-${storyId}`, defaultState);
 }
+
+export function useArgs() {
```
Comment (shilman) [inline lib/api/src/index.tsx]:
```suggestion
export function useArgs(): [Args, (newArgs: Args) => void] {
```

File: lib/api/src/index.tsx [hunk 3]
```diff
@@ -423,3 +427,13 @@ export function useStoryState<S>(defaultState?: S) {
...
+  const {
+    api: { getCurrentStoryData, setStoryArgs },
```
Comment (shilman) [inline lib/api/src/index.tsx]:
```suggestion
    api: { getCurrentStoryData, updateStoryArgs },
```

File: lib/api/src/index.tsx [hunk 4]
```diff
@@ -423,3 +427,13 @@ export function useStoryState<S>(defaultState?: S) {
...
+  } = useStorybookApi();
+
+  const { id, args } = getCurrentStoryData();
+
+  return [args, (newArgs: Args) => setStoryArgs(id, newArgs)];
```
Comment (shilman) [inline lib/api/src/index.tsx]:
```suggestion
  return [args, (newArgs: Args) => updateStoryArgs(id, newArgs)];
```

File: lib/api/src/modules/stories.ts [hunk 5]
```diff
@@ -235,6 +238,24 @@ const initStoriesApi = ({
     }
   };
 
+  const storyArgsChanged = (id: StoryId, args: Args) => {
+    const { storiesHash } = store.getState();
+    (storiesHash[id] as Story).args = args;
+    store.setState({ storiesHash });
+  };
+
+  const setStoryArgs = (id: StoryId, newArgs: Args) => {
```
Comment (shilman) [inline lib/api/src/modules/stories.ts]:
```suggestion
  const updateStoryArgs = (id: StoryId, newArgs: Args) => {
```

File: lib/api/src/tests/stories.test.js [hunk 6]
```diff
@@ -489,6 +512,78 @@ describe('stories API', () => {
...
+
+      setStories({
+        'a--1': { kind: 'a', name: '1', parameters, path: 'a--1', id: 'a--1', args: { a: 'b' } },
+        'b--1': { kind: 'b', name: '1', parameters, path: 'b--1', id: 'b--1', args: { x: 'y' } },
+      });
+
+      const emit = jest.fn();
+      init({ api: { emit, on: jest.fn() } });
+
+      setStoryArgs('a--1', { foo: 'bar' });
```
Comment (shilman) [inline lib/api/src/tests/stories.test.js]:
```suggestion
      updateStoryArgs('a--1', { foo: 'bar' });
```

File: lib/client-api/src/story_store.ts [hunk 7]
```diff
@@ -385,6 +434,14 @@ export default class StoryStore extends EventEmitter {
     this.getStoriesForKind(kind).map(story => this.cleanHooks(story.id));
   }
 
+  setStoryArgs(id: string, newArgs: Args) {
```
Comment (shilman) [inline lib/client-api/src/story_store.ts]:
```suggestion
  updateStoryArgs(id: string, newArgs: Args) {
```

File: lib/core-events/src/index.ts [hunk 8]
```diff
@@ -50,6 +54,8 @@ export const {
   STORY_MISSING,
   STORY_ERRORED,
   STORY_THREW_EXCEPTION,
+  CHANGE_STORY_ARGS,
```
Comment (shilman) [inline lib/core-events/src/index.ts]: `CHANGE` => `UPDATE`?

File: lib/api/src/tests/stories.test.js [hunk 9]
```diff
@@ -489,6 +512,78 @@ describe('stories API', () => {
...
+        kind: 'b',
+        name: '1',
+        parameters: {},
+        path: 'b--1',
+        id: 'b--1',
+        args: { x: 'y' },
+      });
+    });
+
+    it('setStoryArgs emits CHANGE_STORY_ARGS and does not change anything', () => {
```
Comment (shilman) [inline lib/api/src/tests/stories.test.js]:
```suggestion
    it('updateStoryArgs emits CHANGE_STORY_ARGS and does not change anything', () => {
```

File: lib/api/src/tests/stories.test.js [hunk 10]
```diff
@@ -489,6 +512,78 @@ describe('stories API', () => {
...
+      const navigate = jest.fn();
+      const store = createMockStore();
+
+      const {
+        api: { setStories, setStoryArgs },
```
Comment (shilman) [inline lib/api/src/tests/stories.test.js]:
```suggestion
        api: { setStories, updateStoryArgs },
```

File: lib/client-api/src/story_store.test.ts [hunk 11]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      expect(store.getRawStory('a', '1').args).toEqual({
+        arg1: 'arg1',
+        arg2: 3,
+        arg3: { complex: { object: ['type'] } },
+        arg4: 'foo',
+        arg6: false,
+      });
+    });
+
+    it('setStoryArgs changes the args of a story, per-key', () => {
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
    it('updateStoryArgs changes the args of a story, per-key', () => {
```

File: lib/client-api/src/story_store.test.ts [hunk 12]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      const store = new StoryStore({ channel });
+      addStoryToStore(store, 'a', '1', () => 0);
+      expect(store.getRawStory('a', '1').args).toEqual({});
+
+      store.setStoryArgs('a--1', { foo: 'bar' });
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
      store.updateStoryArgs('a--1', { foo: 'bar' });
```

File: lib/client-api/src/story_store.test.ts [hunk 13]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      expect(store.getRawStory('a', '1').args).toEqual({ foo: 'bar' });
+
+      store.setStoryArgs('a--1', { baz: 'bing' });
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
      store.updateStoryArgs('a--1', { baz: 'bing' });
```

File: lib/client-api/src/story_store.test.ts [hunk 14]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      expect(store.getRawStory('a', '1').args).toEqual({ foo: 'bar', baz: 'bing' });
+    });
+
+    it('is passed to the story in the context', () => {
+      const storyFn = jest.fn();
+      const store = new StoryStore({ channel });
+      addStoryToStore(store, 'a', '1', storyFn);
+      store.setStoryArgs('a--1', { foo: 'bar' });
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
      store.updateStoryArgs('a--1', { foo: 'bar' });
```

File: lib/client-api/src/story_store.test.ts [hunk 15]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      store.getRawStory('a', '1').storyFn();
+
+      expect(storyFn).toHaveBeenCalledWith(
+        expect.objectContaining({
+          args: { foo: 'bar' },
+        })
+      );
+    });
+
+    it('setStoryArgs emits STORY_ARGS_CHANGED', () => {
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
    it('updateStoryArgs emits STORY_ARGS_CHANGED', () => {
```

File: lib/client-api/src/story_store.test.ts [hunk 16]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      const onArgsChangedChannel = jest.fn();
+      const testChannel = mockChannel();
+      testChannel.on(Events.STORY_ARGS_CHANGED, onArgsChangedChannel);
+
+      const store = new StoryStore({ channel: testChannel });
+      addStoryToStore(store, 'a', '1', () => 0);
+
+      store.setStoryArgs('a--1', { foo: 'bar' });
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
      store.updateStoryArgs('a--1', { foo: 'bar' });
```

File: lib/client-api/src/story_store.test.ts [hunk 17]
```diff
@@ -84,6 +85,188 @@ describe('preview.story_store', () => {
...
+      expect(onArgsChangedChannel).toHaveBeenCalledWith('a--1', { foo: 'bar' });
+
+      store.setStoryArgs('a--1', { baz: 'bing' });
```
Comment (shilman) [inline lib/client-api/src/story_store.test.ts]:
```suggestion
      store.updateStoryArgs('a--1', { baz: 'bing' });
```

File: lib/client-api/src/story_store.ts [hunk 18]
```diff
@@ -108,6 +111,10 @@ export default class StoryStore extends EventEmitter {
     this._channel.on(Events.SET_CURRENT_STORY, ({ storyId, viewMode }) =>
       this.setSelection({ storyId, viewMode })
     );
+
+    this._channel.on(Events.CHANGE_STORY_ARGS, (id: string, newArgs: Args) =>
+      this.setStoryArgs(id, newArgs)
```
Comment (shilman) [inline lib/client-api/src/story_store.ts]:
```suggestion
      this.updateStoryArgs(id, newArgs)
```

**Review by shilman — APPROVED** [review]
Looks great aside from a few naming issues!

**Pull request merged**